feat: bind mount home strategy

sk1p opened the pull request:
# Change summary

- Add a `BindMount` home strategy, which is useful for example for [snaps](https://kanidm.github.io/kanidm/stable/frequently_asked_questions.html?highlight=snap#why-arent-snaps-launching-with-home_alias-set) on Ubuntu

~~On top of #3985.~~ Rebased on top of current master.

## TODO

This is still WIP, and needs:

- [ ] Edge cases, like how to handle things that exist at the target path and are not a directory (or symlink)? Log an error and continue?
- [ ] Support for non-Linux systems? I've added dependencies on `procfs` and `nix`, and I'm not sure if that aligns with the supported platforms
- [ ] Documentation updates
- [ ] Changes in the systemd services sandboxing - the current settings prevent the mount syscall from working because of capabilities and syscall filtering, and hides the mounts from the hosts system by mount namespacing. It might make sense to have different sandboxing, to keep the security level of the default configuration as high as possible
- [ ] Test strategy - are tests executed as root somewhere in the CI pipeline?

Fixes #3970 

Checklist

- [x] This PR contains no AI generated code
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

/cc @ReimarBauer @Firstyear

sk1p committed:
WIP: add BindMount HomeStrategy

sk1p commented:
As a comment on current status: @ReimarBauer has made test deployments on both Debian and Ubuntu test systems, and the general idea seems to work. We are still trying to figure out the systemd-related changes needed, especially exactly which sandboxing flags can stay enabled in this configuration.

Firstyear reviewed (changes requested):
Thanks, it's a good start, some feedback inline :)

Firstyear commented on unix_integration/resolver/src/bin/kanidm_unixd_tasks.rs:
```diff
@@ -185,98 +202,178 @@ fn create_home_directory(
-                    return Err(format!("{e:?}"));
-                }
+    // Finally, try to create the bind mount
+    nix::mount::mount::<Path, Path, str, str>(
```
ahhh! That certainly clears that up then!

sk1p committed:
Refine handling of existing mounts

If a mount exists at the alias path which does not point to the correct actual
home directory, umount it. Adjust "already mounted" branch to check the mount
source path.

Firstyear commented on unix_integration/resolver/src/bin/kanidm_unixd_tasks.rs:
```diff
@@ -22,14 +22,17 @@ use kanidm_unix_common::unix_proto::{
 };
 use kanidm_utils_users::{get_effective_gid, get_effective_uid};
 use libc::{lchown, umask};
+use nix::mount::MsFlags;
```
```
error[E0433]: failed to resolve: use of unresolved module or unlinked crate `nix`
  --> unix_integration/resolver/src/bin/kanidm_unixd_tasks.rs:25:5
   |
25 | use nix::mount::MsFlags;
   |     ^^^ use of unresolved module or unlinked crate `nix`
   |
   = help: if you wanted to use a crate named `nix`, use `cargo add nix` to add it to your `Cargo.toml`

error[E0433]: failed to resolve: use of unresolved module or unlinked crate `procfs`
  --> unix_integration/resolver/src/bin/kanidm_unixd_tasks.rs:30:5
   |
30 | use procfs::process::Process;
   |     ^^^^^^ use of unresolved module or unlinked crate `procfs`
   |
   = help: if you wanted to use a crate named `procfs`, use `cargo add procfs` to add it to your `Cargo.toml`

error[E0433]: failed to resolve: use of unresolved module or unlinked crate `nix`
   --> unix_integration/resolver/src/bin/kanidm_unixd_tasks.rs:261:9
    |
261 |         nix::mount::umount(&m.mount_point).map_err(|e| {
    |         ^^^ use of unresolved module or unlinked crate `nix`
    |
    = help: if you wanted to use a crate named `nix`, use `cargo add nix` to add it to your `Cargo.toml`

error[E0433]: failed to resolve: use of unresolved module or unlinked crate `nix`
   --> unix_integration/resolver/src/bin/kanidm_unixd_tasks.rs:278:5
    |
278 |     nix::mount::mount::<Path, Path, str, str>(
    |     ^^^ use of unresolved module or unlinked crate `nix`
    |
    = help: if you wanted to use a crate named `nix`, use `cargo add nix` to add it to your `Cargo.toml`

For more information about this error, try `rustc --explain E0433`.
error: could not compile `kanidm_unix_int` (bin "kanidm_unixd_tasks") due to 4 previous errors
```

Errors on non-linux platforms, so I think we need to cfg flag this.

Firstyear commented on unix_integration/resolver/src/bin/kanidm_unixd_tasks.rs:
```diff
@@ -222,9 +225,70 @@ fn create_home_directory(
 
+fn home_alias_update_bind_mount(alias_path: &Path, hd_mount_path: &Path) -> Result<(), String> {
+    // If the alias_path is a symlink, remove it
+    // TODO: what if it exsits and is not a symlink?
```
Then we don't touch it, because I don't want to risk damage to someones data. It should be up to a human to intervene.

yaleman replied:
Should probably throw a warning on login when bind mount's enabled and they find a symlink or existing dir though...

Firstyear reviewed (commented):
Okay this is looking pretty good.

Sadly on the testing front, we do these tests a bit more manually, maybe we shuold look at some CI, or something "else" to make a framework here that we can use to test. Depends how far into this we want to go. We could do an abstraction over the mount / file tools, and then make a trait for them that we can shim out/in and have a mock version. 

Either way this is looking great, I think it's just some polish and testing and we're there. I'll give this a go later on a linux system.

sk1p commented:
> Sadly on the testing front, we do these tests a bit more manually, maybe we shuold look at some CI, or something "else" to make a framework here that we can use to test. Depends how far into this we want to go. 

Okay. I tried to set up a manual test environment before Christmas but ran out of time. I think I didn't yet fully nail the configuration, as there was somehow a mismatch between the directory used for $HOME and the path sent to the kanidm UNIX daemon. I hope to work on it again next week.

> We could do an abstraction over the mount / file tools, and then make a trait for them that we can shim out/in and have a mock version.

Indeed, a trait for a mount backend of some kind is how I implemented this for a different project. I'll think a bit if I can come up with a "quick win" on the CI side of things. 

> Either way this is looking great, I think it's just some polish and testing and we're there. I'll give this a go later on a linux system.

Sounds good! For deployment, we/I still need to nail down the systemd service configuration, especially with regards to sandboxing etc. Do you prefer to have multiple smaller PRs with changes to documentation and deployment/service definitions, or should this be part of this PR?

Also, is it OK if I rebase my branch, or do you prefer merge commits? Thanks!

yaleman commented:
Definitely rebase, we squash on merge anyway.

sk1p committed:
Add `home_strategy` to `Display` of `UnixdConfig`

And a little drive-by typo fix.

sk1p commented:
> Could you rebase / update this PR so we can consider it for merge? Or would you like us to finish it up?

I rebased, but please feel free to take it over, as I still didn't have time to set up a proper test environment or continue with @reimarbauer 's deployment. Thanks!

ReimarBauer commented:
We're currently using a test environment based on Debian and Ubuntu (2 machines) within our institute’s systems, built from a previous iteration of this project. This institute is separate from @sk1p’s area of work.  I don't have anyone on my team who can assist with RUST, so I really appreciate @sk1p taking the initiative. We'd much rather deploy an official package than apply the patch manually.
I think this option should be configurable, not the default. The existing solution is a good starting point for others, so I'd appreciate it if you @yaleman @Firstyear could take over and merge this.

Firstyear commented:
No problem, I'll rebase and wrap this up for you soon then.

Firstyear committed:
Properly cfg define other oses

Firstyear committed:
Work out the needed systemd flags

Firstyear commented:
Okay, I worked out the needed systemd settings to allow this. It seems that most of systemds protections interfer with "mounting" so as a result, we have to disabled pretty much all their hardening. Disappointing, but that's how it is.

Firstyear committed:
lints

yaleman commented on unix_integration/resolver/debian/kanidm-unixd-tasks.service:
we should document the different requirements based on what bind mount strategy is being used. Some users would prefer the service not have `CAP_SYS_ADMIN` and use Symlink instead.

yaleman commented on unix_integration/resolver/src/bin/kanidm_unixd_tasks.rs:
```diff
@@ -222,7 +220,85 @@ fn create_home_directory(
+    // Remove conflicting mount if it exists:
+    let mismatching_mount = current_mounts.iter().find(|m| {
+        m.mount_point == alias_path
+            && m.mount_source != Some(hd_mount_path.to_string_lossy().to_string())
```
I feel like we should be pushing m.mount_source to be a Path instead of the messy `.to_string_lossy().to_string()`

Firstyear committed:
Apply suggestions from code review

Co-authored-by: James Hodgkinson <james@terminaloutcomes.com>

Firstyear committed:
Update based on review feedback

Firstyear committed:
Update

ReimarBauer commented on examples/unixd:
```diff
@@ -93,6 +94,24 @@ version = '2'
+#
+# (linux only)
+# * bind_mount: bind mounts the alias directory to the stable path. This requires modification of
+#               .service file on linux to grant the needed permissions. For more details, see:
```
@Firstyear `home_strategy = "bindmount"`without the `_` works for me. 

That gets readed without an error /usr/sbin/kanidm_unixd  -t -r

```
...
home_strategy: BindMount
...
```

ReimarBauer reviewed (commented):
just seen that there is a typo in the the word bindmount
